README.md

# Ansible etckeeper actions

## Introduction

[Ansible](http://ansible.com) is a tool to automate configuration and management of one or more hosts. [Etckeeper](https://etckeeper.branchable.com/) is a tool to put file changes under the `/etc` directory into version control.

This repository contrains Ansible actions so that when an Ansible task changes something
under the `/etc` directory, those changes are checked in into version control.

## Description

If you install a package with the
[package](https://docs.ansible.com/ansible/latest/modules/package_module.html) module
(or
[apt](https://docs.ansible.com/ansible/latest/modules/apt_module.html),
[yum](https://docs.ansible.com/ansible/latest/modules/yum_module.html)
or
[dnf](https://docs.ansible.com/ansible/latest/modules/dnf_module.html))
then
[etckeeper](https://etckeeper.branchable.com/)
will be automatically called and changes in `/etc` will be committed.
However changes caused by other modules like for instance
[lineinfile](https://docs.ansible.com/ansible/latest/modules/lineinfile_module.html)
or
[user](https://docs.ansible.com/ansible/latest/modules/user_module.html)
are not committed automatically.

With the three actions `etckeeper-pre-task`, `etckeeper-commit-task`, and
`etckeeper-post-task` you can make sure that any changes in `/etc` triggered
by an ansible task are commited.


## Example Playbook

Here is an example of how to capture changes made by the `user` module:

```yaml
---
- hosts: all
  gather_facts: no
  tasks:
    - etckeeper-pre-task:

    - name: Remove user Darl McBribe
      user:
        name: dmcbribe
        state: absent
        remove: yes
      register: user_result

    - etckeeper-post-task:
      when: user_result.changed
```

If there were uncommitted changes when the playbook started, those were checked in
with commit message `saving uncommitted changes in /etc prior to ansible task run`.

If the user were removed, that change would be checked in with a commit with message
`saving uncommitted changes in /etc prior to ansible task run`.

To set a custom message on a commit:
```yaml
---
- hosts: all
  gather_facts: no
  tasks:
    - etckeeper-commit-task:
      msg: Etckeeper commit in role_name taskfile before some_change
```


## Dependencies

Etckeeper must be installed and initialized. Notice however that if etckeeper is not
installed on a host, the actions does not fail so it is safe to add them to
a playbook even if not all hosts use etckeeper.

## Requirements

The actions need to be run as root.

## Installation

TBD

## Configuration

There is nothing to configure for the actions.

## Contributing

You are welcome to create pull requests if you think there is something that can be
improved.

### TODO list

* Figure out how to share code between the three files.

## FAQ

### Why not using `notify` and `handlers` instead of `actions`?

Because handler invocations are delayed to the end, multiple calls are joined and
reduced to one, and handlers might not be called under some circumstances.

### But I do not want to pepper my playbooks with calls to etckeeper actions around all tasks that can modify `/etc`.

You should and you have to. If you think that you rather would have one commit for
all changes made by a playbook, you are forgetting that any calls to `package`
[et al.](https://en.wiktionary.org/wiki/et_al.#English) will create individual
commits inbetween thus ruining that idea.

It is possible to achieve this idea if you first check out a new *playbook run* branch,
create commits on that one and then at the end do a non-fast-forward merge back to master.
But you still should create individual commit for individual changes,
which is good version control practice in any case.

Now if you really want to put in minium effort, it is possible to put just one call
to etckeeper-pre-task at the very beginning and one unconditional call to
etckeeper-post-task at the very end. That will work fine, just not giving
you the full granularity you could have had.

## Version

The latest version is 1.0.1. See the [changelog](./CHANGELOG.md) for details.

## Maintainers

Håkon Løvdal <kode@denkule.no>

## License

[GPL](LICENSE.txt) version 3 or (at your option) any later version. ([summary](https://tldrlegal.com/l/gpl-3.0))


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Security issues

### Added

### Changed

### Removed

### Fixed

### Breaking changes

## [1.0.1] - 2023-12-06

### Added

- Added an additional etckeeper action that allows for setting custom msg.

### Changed

- Renamed `master` branch to `main`.
- Make check for "nothing to commit" state less strict to support older versions
  of etckeeper (CentOS7 EPEL7's version).

## [1.0.0] - 2018-12-09

### Added

- First release.


action_plugins/etckeeper-post-task.py
#!/usr/bin/python

# Copyright: (c) 2018, Håkon Løvdal <kode@denkule.no>
# GNU General Public License v3.0+ (see gpl.txt, https://tldrlegal.com/l/gpl-3.0 or https://www.gnu.org/licenses/gpl-3.0.txt)

# TODO: Find a way to share this between the files. Failed attempts:
#from lib_etckeeper import run_etckeeper   # ERROR! Unexpected Exception, this is probably a bug: No module named 'lib_etckeeper'
#from .lib_etckeeper import run_etckeeper  # ERROR! Unexpected Exception, this is probably a bug: No module named 'ansible.plugins.action.lib_etckeeper'
#from . import lib_etckeeper               # ERROR! Unexpected Exception, this is probably a bug: cannot import name 'lib_etckeeper' from 'ansible.plugins.action' (/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py)

import os

def run_etckeeper(self, msg):
    # Make sure we are not parsing localized text
    os.environ["LC_ALL"] = "C"

    result = self._low_level_execute_command('etckeeper commit "%s"' % msg)

    # In case of no changes the command will fail, but ignore that
    if (result['rc'] == 1) and ('nothing to commit' in result['stdout']):
        result['rc'] = 0

    # If etckeeper is not installed, do not treat that as an error.
    # Return code 127 is established behaviour for command not found. http://www.tldp.org/LDP/abs/html/exitcodes.html, https://stackoverflow.com/q/1763156/23118
    if (result['rc'] == 127) and ('command not found' in result['stdout']):
        result['rc'] = 0

    return result


from ansible.plugins.action import ActionBase

class ActionModule(ActionBase):

    TRANSFERS_FILES = False

    def run(self, tmp=None, task_vars=None):
        msg = 'committing changes in /etc after ansible task run'
        result = run_etckeeper(self, msg)
        return result


action_plugins/etckeeper-commit-task.py
#!/usr/bin/python

# Copyright: (c) 2018, Håkon Løvdal <kode@denkule.no>
# GNU General Public License v3.0+ (see gpl.txt, https://tldrlegal.com/l/gpl-3.0 or https://www.gnu.org/licenses/gpl-3.0.txt)

# TODO: Find a way to share this between the files. Failed attempts:
#from lib_etckeeper import run_etckeeper   # ERROR! Unexpected Exception, this is probably a bug: No module named 'lib_etckeeper'
#from .lib_etckeeper import run_etckeeper  # ERROR! Unexpected Exception, this is probably a bug: No module named 'ansible.plugins.action.lib_etckeeper'
#from . import lib_etckeeper               # ERROR! Unexpected Exception, this is probably a bug: cannot import name 'lib_etckeeper' from 'ansible.plugins.action' (/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py)

import os

def run_etckeeper(self, msg):
    # Make sure we are not parsing localized text
    os.environ["LC_ALL"] = "C"

    result = self._low_level_execute_command('etckeeper commit "%s"' % msg)

    # In case of no changes the command will fail, but ignore that
    if (result['rc'] == 1) and ('nothing to commit' in result['stdout']):
        result['rc'] = 0

    # If etckeeper is not installed, do not treat that as an error.
    # Return code 127 is established behaviour for command not found. http://www.tldp.org/LDP/abs/html/exitcodes.html, https://stackoverflow.com/q/1763156/23118
    if (result['rc'] == 127) and ('command not found' in result['stdout']):
        result['rc'] = 0

    return result


from ansible.plugins.action import ActionBase

class ActionModule(ActionBase):

    TRANSFERS_FILES = False

    def run(self, tmp=None, task_vars=None):

        validation_result, new_module_args = self.validate_argument_spec(
            argument_spec={
                'msg': {'type': 'raw', 'default': 'saving uncommitted changes in /etc within ansible run'},
            },
        )

        msg = new_module_args['msg']

        result = run_etckeeper(self, msg)
        return result


action_plugins/etckeeper-pre-task.py
#!/usr/bin/python

# Copyright: (c) 2018, Håkon Løvdal <kode@denkule.no>
# GNU General Public License v3.0+ (see gpl.txt, https://tldrlegal.com/l/gpl-3.0 or https://www.gnu.org/licenses/gpl-3.0.txt)

# TODO: Find a way to share this between the files. Failed attempts:
#from lib_etckeeper import run_etckeeper   # ERROR! Unexpected Exception, this is probably a bug: No module named 'lib_etckeeper'
#from .lib_etckeeper import run_etckeeper  # ERROR! Unexpected Exception, this is probably a bug: No module named 'ansible.plugins.action.lib_etckeeper'
#from . import lib_etckeeper               # ERROR! Unexpected Exception, this is probably a bug: cannot import name 'lib_etckeeper' from 'ansible.plugins.action' (/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py)

import os

def run_etckeeper(self, msg):
    # Make sure we are not parsing localized text
    os.environ["LC_ALL"] = "C"

    result = self._low_level_execute_command('etckeeper commit "%s"' % msg)

    # In case of no changes the command will fail, but ignore that
    if (result['rc'] == 1) and ('nothing to commit' in result['stdout']):
        result['rc'] = 0

    # If etckeeper is not installed, do not treat that as an error.
    # Return code 127 is established behaviour for command not found. http://www.tldp.org/LDP/abs/html/exitcodes.html, https://stackoverflow.com/q/1763156/23118
    if (result['rc'] == 127) and ('command not found' in result['stdout']):
        result['rc'] = 0

    return result


from ansible.plugins.action import ActionBase

class ActionModule(ActionBase):

    TRANSFERS_FILES = False

    def run(self, tmp=None, task_vars=None):
        msg = 'saving uncommitted changes in /etc prior to ansible task run'
        result = run_etckeeper(self, msg)
        return result
